### PR TITLE
Allow for overriding build tools via the command line or environment

### PR DESCRIPTION
**Pull request by ymanton**
These patches allow for building OpenJ9 with alternative build tools by overriding `CC`, `CXX`, and the other standard GNU make implicit variables, as well as OpenJ9-specific variables.

Note: OMR still has issues and will require build tools known to autoconf to be in your path, and overriding `CC` and `CXX` with an absolute path will break OpenJDK's `configure`, so use just the tool name, e.g. `CC=gcc-4.8` and make sure it's in your path.

Fixes: #1180

### CODE REVIEW

**Comment by DanHeidinga** [comment]
@ymanton Any suggestions for reviewers for the JIT makefiles?  Anyone particularly knowledgable in that area who can look would be appreciated.

**Comment by ymanton** [comment]
@JamesKingdon, you've poked around the JIT makefiles recently for ARM. Can you take a look?

@fjeremic may have also poked around those parts when he re-organized the compiler sub-tree.

File: runtime/makelib/mkconstants.mk.ftl [hunk 1]
```diff
@@ -33,27 +33,70 @@ ifndef UMA_TARGET_PATH
...
-<#if uma.spec.tools.ar.needed>AR=${uma.spec.tools.ar.name}<#else># AR not used</#if>
-<#if uma.spec.tools.ranlib.needed>RANLIB=${uma.spec.tools.ranlib.name}<#else># RANLIB not used</#if>
-<#if uma.spec.tools.dll_ld.needed>DLL_LD=${uma.spec.tools.dll_ld.name}<#else># DLL_LD not used</#if>
-<#if uma.spec.tools.cxx_dll_ld.needed>CXX_DLL_LD=${uma.spec.tools.cxx_dll_ld.name}<#else># CXX_DLL_LD not used</#if>
-<#if uma.spec.tools.exe_ld.needed>EXE_LD=${uma.spec.tools.exe_ld.name}<#else># EXE_LD not used</#if>
-<#if uma.spec.tools.cxx_exe_ld.needed>CXX_EXE_LD=${uma.spec.tools.cxx_exe_ld.name}<#else># CXX_EXE_LD not used</#if>
-<#if uma.spec.tools.dll_ld.needed>UMA_DLL_LD=$(if $(UMA_IS_C_PLUS_PLUS),$(CXX_DLL_LD),$(DLL_LD))</#if>
-<#if uma.spec.tools.exe_ld.needed>UMA_EXE_LD=$(if $(UMA_IS_C_PLUS_PLUS),$(CXX_EXE_LD),$(EXE_LD))</#if>
+<#if uma.spec.tools.interp_gcc.needed>
+# Order of preference: INTERP_GCC if set, CC if set and not default, otherwise uma.spec.tools.interp_gcc.name
```
Comment (DanHeidinga) [inline runtime/makelib/mkconstants.mk.ftl]: Calling this out for whoever reviews this - the order here needs some careful thought as I'm not sure if the CC before uma.spec.tools.interp_gcc.name is correct

**Comment by JamesKingdon** [comment]
Looks good to me, although I surely can't tell from looking at it if it's going to work or not. It's clearly a major improvement, so let's put it in and fix any problems as they arise. Thanks for tackling this painful task.

File: runtime/makelib/mkconstants.mk.ftl (hunk 1 above)
Comment (DanHeidinga) [reply runtime/makelib/mkconstants.mk.ftl]: Extensions repos may be hacking around this differently - see https://github.com/ibmruntimes/openj9-openjdk-jdk8/blob/84153c75da79342522533371b19330181ab4ff4f/closed/OpenJ9.gmk#L227

Comment (ymanton) [reply runtime/makelib/mkconstants.mk.ftl]: My thinking here is that if a typical user sets `CC` the expectation will be that that everything is built with the requested compiler. If we don't allow `CC` to set `INTERP_GCC` they'll have to dig around to discover what's going on and override it themselves, which is an annoyance I wouldn't want to inflict.

On the flip side anyone who's aware of `INTERP_GCC` can still set it themselves if they override `CC`, although it just occurred to me that they'd have to know what the hardcoded value was, so perhaps another var like `INTERP_GCC_IGNORE_CC` to control this behaviour makes sense, that way you can set your own value or keep the hardcoded value without knowing/caring what it is.

**Comment by DanHeidinga** [comment]
I've tested this PR on an Ubuntu 16 machine with `CC` and `CXX` set and was able to successfully build an OpenJ9 JDK.

```
export CC=gcc-4.8
export CXX=g++-4.8
```

**Comment by DanHeidinga** [comment]
Jenkins test sanity

**Comment by ymanton** [comment]
I decided to overrule myself on the `INTERP_GCC` bit because there's no guarantee that `CC` and `INTERP_GCC` are interchangeable; for example if `CC` is XLC or MSVC and the user sets it to an alternative XLC or MSVC it wouldn't make sense to set `INTERP_GCC` to either of those and would break the build.

In the interest of giving the user an idea of what's going on I've changed that part to print a notice if `CC` has been overridden but `INTERP_GCC` has not; it looks like the following and pauses for 5 seconds (to make it easier to spot):

```
****************
*
* CC=/usr/bin/gcc-4.8 (overridden), note that this build will also invoke another compiler that can be overridden: INTERP_GCC=gcc
*
****************
```

There are also `MINGW_CXX` and `PPC_GCC_CXX`, which are similar to `INTERP_GCC` which I missed earlier; these are handled same as above now, except they look for `CXX` being overridden.

File: runtime/makelib/mkconstants.mk.ftl [hunk 2]
```diff
@@ -33,27 +33,137 @@ ifndef UMA_TARGET_PATH
...
+ifneq (default,$(origin CC))
+  ifndef INTERP_GCC
+    # If the user has overridden CC we'll want to let them know that INTERP_GCC exists.
+    ifndef PRINT_ONCE_INTERP_GCC
+      $(info ****************)
+      $(info *)
+      $(info * CC=$(CC) (overridden), note that this build will also invoke another compiler that can be overridden: INTERP_GCC=${uma.spec.tools.interp_gcc.name})
+      $(info *)
+      $(info ****************)
+      $(shell sleep 5)
```
[the remark this answers is not in the document]
Comment (ymanton) [reply runtime/makelib/mkconstants.mk.ftl]: Sleep removed.

**Comment by DanHeidinga** [comment]
Jenkins test sanity

**Comment by DanHeidinga** [comment]
The JDK9 builds all failed due to freemarker.jar issues (see fix in pr #1406).

Relaunching just the 9 builds

**Comment by DanHeidinga** [comment]
Jenkins test sanity jdk9

**Comment by DanHeidinga** [comment]
Jenkins test sanity all jdk9

**Pull request merged**